nav2_waypoint_follower; Add an optional param to specify a sleep time…

jediofgever opened the pull request:
Signed-off-by: jediofgever <fetulahatas1@gmail.com>

<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (https://github.com/ros-planning/navigation2/issues/1992) |
| Primary OS tested on | (Ubuntu) |
| Robotic platform tested on | (Gazebo simulation of Turtlebot3) |

---

## Description of contribution in a few bullet points

 
* Added an additional/(optional) int parameter to `nav2_waypoint_follower`, in order to control behavior of robot in between waypoints. This int parameter species an amount of time in seconds to wait/sleep after reaching each waypoint. The parameter could be set to zero if a continuous execution(or as in original implementation) of waypoints is desired.  
 

## Description of documentation updates required from your changes

 
* Added new parameter, so need to add that to default configs and documentation page on doc

 

---

## Future work that may be required in bullet points
 
* A more advanced concept for controlling robot's behavior on arrival of waypoints, could be based on service/action as mentioned by @SteveMacenski in https://github.com/ros-planning/navigation2/issues/1992#issuecomment-696955758 , However I might open another ticker for that later on.

jediofgever commented:
All the checks have failed but I don't understand why ? I targeted `foxy-devel` branch and did some very minor changes.

fmrico commented:
Hi @jediofgever Did you passed tests? I think that some code would not pass the linters...

jediofgever commented:
Hi @fmrico , I haven't really run tests. Could you elaborate more on how we can run tests locally ?

SteveMacenski commented:
@jediofgever you must target the `main` branch for any pull requests.

SteveMacenski commented on doc/parameters/param_list.md:
```diff
@@ -479,6 +479,8 @@ When `planner_plugins` parameter is not overridden, the following default plugin
 | ----------| --------| ------------|
 | stop_on_failure | true | Whether to fail action task if a single waypoint fails. If false, will continue to next waypoint. |
 | loop_rate | 20 | Rate to check for results from current navigation task |
+| sleep_time_inbetween_waypoints | 5 | Amount of time in seconds for robot to sleep/wait after each waypoint is reached. If zero, robot will directly continue to next waypoint. |
```
Also must be added to the waypoint follower entry on navigation.ros.org + migration guides https://github.com/ros-planning/navigation.ros.org/blob/master/configuration/packages/configuring-waypoint-follower.rst

SteveMacenski replied:
Default must also be `0` to retain same behavior.

SteveMacenski commented on nav2_bringup/bringup/params/nav2_params.yaml:
```diff
@@ -279,3 +279,9 @@ recoveries_server:
+  ros__parameters:
+    loop_rate: 20
+    stop_on_failure: false   
+    sleep_time_inbetween_waypoints: 5    
```
Default 0

SteveMacenski replied:
and add new line

SteveMacenski commented on nav2_waypoint_follower/src/waypoint_follower.cpp:
```diff
@@ -31,6 +31,7 @@ WaypointFollower::WaypointFollower()
 
   declare_parameter("stop_on_failure", true);
   declare_parameter("loop_rate", 20);
+  declare_parameter("sleep_time_inbetween_waypoints", 5);
```
Default 0

SteveMacenski commented on nav2_waypoint_follower/src/waypoint_follower.cpp:
```diff
@@ -45,6 +46,7 @@ WaypointFollower::on_configure(const rclcpp_lifecycle::State & /*state*/)
 
   stop_on_failure_ = get_parameter("stop_on_failure").as_bool();
   loop_rate_ = get_parameter("loop_rate").as_int();
+  sleep_time_inbetween_waypoints_ = get_parameter("sleep_time_inbetween_waypoints").as_int();
```
This should be a double, not an int

jediofgever replied:
I also thought of double at first, but using double with `time` is troublesome and often will need a casting anyway, I suggest we stick with int, though we can change the  unit to milliseconds ?

SteveMacenski replied:
I don’t think it matters. Up to you. Int is OK for ms, but double for seconds.

SteveMacenski commented on nav2_waypoint_follower/src/waypoint_follower.cpp:
```diff
@@ -182,9 +184,17 @@ WaypointFollower::followWaypoints()
-      RCLCPP_INFO(
-        get_logger(), "Succeeded processing waypoint %i, "
-        "moving to next.", goal_index);
+        if (sleep_time_inbetween_waypoints_){
```
This definitely has linting errors, please run the linters on your PR

SteveMacenski reviewed (changes requested):
Another option I was thinking over this afternoon was adding a plugin for processing. 

So rather than just hardcoding the sleep, we could create a really simple pluginlib interface for the user to load a plugin to do some behavior.

For example:
- Plugin has an `initialize()` and a `run()` function. Initialize will let the plugin setup any action / service / get parameters / etc for use.
- Then once we reach a waypoint, we give that waypoint pose to the `run()` function which can do whatever it likes. For this implementation, that plugin would just sleep for `N` seconds
- For a more specific use-case (like taking a picture or waiting for external input) then it could call a service, action, or wait for an event to occur before continuing on.

This would allow for all of those types of behaviors and abstract out the specifics to a run-time loadable plugin.

jediofgever commented:
Applied requested changes and , changed trageted branch in this commit https://github.com/jediofgever/navigation2/commit/2a6d1b17b0d93de477d8219f5898ab639e4dbf31

I opened a new PR https://github.com/ros-planning/navigation2/pull/1996 with modified changes but CI builds are failing, interestingly on the package that I modified. 
https://app.circleci.com/pipelines/github/ros-planning/navigation2/3798/workflows/32670729-f667-4f8c-8244-afe68822981a/jobs/14762

but from the log I see the error has appeared before I submitted https://github.com/ros-planning/navigation2/pull/1996

I closed the that PR, I will open once the CI is back to normal.

jediofgever commented:
> Another option I was thinking over this afternoon was adding a plugin for processing.
> 
> So rather than just hardcoding the sleep, we could create a really simple pluginlib interface for the user to load a plugin to do some behavior.


processing this information  ...
We agree that Each user's desired task to be executed on waypoint arrival could be different. 
Therefore It is quite hard to think of a general purpose action server. 

In my case, at each waypoint arrival, I had like call a action server which will notify user with QT pop-up, depending on user's reaction, the robot will then proceed for the next way point. 

I was thinking of adding a generic action such as ; 

`ExecuteTaskAtWaypoint.action`
```
# Goal
string task_name
geometry_msgs/PoseStamped pose
---
# Result
builtin_interfaces/Duration total_elapsed_time
---
# Feedback
float32 percent_complete 
```
Then again the implementation of the server for side for this action will be depending on developer's need. 

Now back to your comment; 

> So rather than just hardcoding the sleep, we could create a really simple pluginlib interface for the user to load a plugin to do   some behavior.

I think that if it is decided that we add this `ExecuteTaskAtWaypoint.action`, this plugin will be basically a wrapper for the client side of purposed action, am I right ?

SteveMacenski commented:
Please do not open new PRs for the same topic. Add your commit to this branch / PR for review. Submitting a new PR for each new commit makes it impossible to track changes and spams subscribers unnecessarily

fmrico commented:
> Hi @fmrico , I haven't really run tests. Could you elaborate more on how we can run tests locally ?

Hi @jediofgever

To pass the tests, do a `colcon test`  and check the log of the packages that didn't pass it. Tests include linters. In log dir, in latest_test, each package has a folder with stdout and stderr of this process. You can use the `--packages-select` option.

I checked in the CI details and I saw some style errors, like lines with spaces and `{` in the wrong line.

jediofgever commented:
> Please do not open new PRs for the same topic. Add your commit to this branch / PR for review. Submitting a new PR for each new commit makes it impossible to track changes and spams subscribers unnecessarily
my bad , I opened that PR because I targeted the wrong branch at first place, lets assume that PR never happened,I will add modified code into this PR.

SteveMacenski commented on doc/parameters/param_list.md:
```diff
@@ -479,8 +479,7 @@ When `planner_plugins` parameter is not overridden, the following default plugin
 | loop_rate | 20 | Rate to check for results from current navigation task |
-| sleep_time_inbetween_waypoints | 5 | Amount of time in seconds for robot to sleep/wait after each waypoint is reached. If zero, robot will directly continue to next waypoint. |
-
+| sleep_time_inbetween_waypoints | 0 | Amount of time in milliseconds, for robot to sleep/wait after each waypoint is reached. If zero, robot will directly continue to next waypoint. |
```
`waypoint_pause_duration` is a little more descriptive and shorter

SteveMacenski replied:
and variable name / website update

SteveMacenski commented:
Yes, a simple plugin definition in `/include` with implementations in `/plugins` would work. I think the API should be something like having an initialization function taking in a name and pointer to node. Then a `processAtWaypoint()` function taking in a pose & waypoint number to do whatever it likes (in this case, just sleeping).

jediofgever commented:
Applied the suggestions to the draft. I think the above commit is ready for a review. I hope to iterate through,  in this week, I will take long break and might not be back quite soon. 
In `processAtWaypoint()` we have provided all we can actually :) 

```
void WaitAtWaypointArrival::processAtWaypoint(
  const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
  const geometry_msgs::msg::PoseStamped curr_pose, const int curr_waypoint_index);

```
in addition to that I guess I would be able to add the parameter entries to https://github.com/ros-planning/navigation.ros.org 
soon.

SteveMacenski commented:
Yes, for docs adding parameters to this repo's yaml and then the website's waypoint follower page. Additionally adding this to the website's migration guides. At some point (not part of this PR but hopefully if you're willing) a tutorial about how to create a custom plugin for their applications.

SteveMacenski commented on doc/parameters/param_list.md:
```diff
@@ -480,6 +480,13 @@ When `planner_plugins` parameter is not overridden, the following default plugin
 | stop_on_failure | true | Whether to fail action task if a single waypoint fails. If false, will continue to next waypoint. |
 | loop_rate | 20 | Rate to check for results from current navigation task |
 
+## task_executor_at_waypoint plugin
```
I think in the wp follower entry you need to add the `plugin` parameter that loads this one.

SteveMacenski commented on doc/parameters/param_list.md:
```diff
@@ -480,6 +480,13 @@ When `planner_plugins` parameter is not overridden, the following default plugin
+
+* `<task executor at waypoint>`: Name corresponding to the `nav2_waypoint_follower::WaitAtWaypointArrival` plugin. 
+ 
+| `<task executor at waypoint>`.enabled | false | Whether it is enabled |
```
Why default disabled? That doesn't seem right.

SteveMacenski commented on nav2_bringup/bringup/params/nav2_params.yaml:
```diff
@@ -279,3 +279,13 @@ recoveries_server:
+    task_executor_at_waypoint_plugin: "task_executor_at_waypoint"   
+    task_executor_at_waypoint:
+      plugin: "nav2_waypoint_follower::WaitAtWaypointArrival"
+      enabled: False
```
True?

SteveMacenski commented on nav2_core/include/nav2_core/task_executor_at_waypoint_arrival.hpp:
```diff
@@ -0,0 +1,88 @@
+ * @brief Base class for creating a plugin in order to perform a specific task at waypoint arrivals.
+ *
+ */
+class TaskExecutorAtWaypointArrival
```
I think this name is too long, we need to shorten it. `TaskExecutor` sounds good, but I agree some reference to waypoint or pose would be good. `WaypointTaskExecutor`?

SteveMacenski commented on nav2_core/include/nav2_core/task_executor_at_waypoint_arrival.hpp:
```diff
@@ -0,0 +1,88 @@
+   * @param curr_waypoint_index current waypoint, that robot just arrived
+   */
+  virtual void processAtWaypoint(
+    const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
```
You don't need to give the parent again, they should have stored it if they wanted it

SteveMacenski replied:
Make the other 2 things const references

SteveMacenski commented on nav2_waypoint_follower/include/nav2_waypoint_follower/plugins/wait_at_waypoint_arrival.hpp:
```diff
@@ -0,0 +1,98 @@
+/*
+ * Software License Agreement (BSD License)
```
Please change the licenses to Apache 2.0. The package is all apache 2.0 so if you made modifications to it, please use the same license

SteveMacenski replied:
As well in all files

SteveMacenski commented on nav2_waypoint_follower/include/nav2_waypoint_follower/plugins/wait_at_waypoint_arrival.hpp:
```diff
@@ -0,0 +1,98 @@
+ *        your own task and rewrite the body for it.
+ *
+ */
+class WaitAtWaypointArrival : public nav2_core::TaskExecutorAtWaypointArrival
```
How do you feel about `WaitAtWaypoint` ?

SteveMacenski commented on nav2_waypoint_follower/plugins/wait_at_waypoint_arrival.cpp:
```diff
@@ -0,0 +1,100 @@
+  const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
+  const std::string & plugin_name)
+{
+  auto node = parent.lock();
```
Do a `if (!node) {throw std::runtime_exception("some-error-node-not-valid")}` after you lock to make sure its valid

SteveMacenski commented on nav2_waypoint_follower/plugins/wait_at_waypoint_arrival.cpp:
```diff
@@ -0,0 +1,100 @@
+    return;
+  }
+
+  auto node = parent.lock();
```
In `initialize()` store the `logger_` as a class variable so you don't have to lock this at runtime when processing.

SteveMacenski commented on nav2_waypoint_follower/plugins/wait_at_waypoint_arrival.cpp:
```diff
@@ -0,0 +1,100 @@
+  if (!is_enabled_) {
+    RCLCPP_INFO(
+      node->get_logger(), "Waypoint task executor plugin is disabled.");
+  }
```
I'd also check if duration is 0, if so, set the is_enabled_ to false since it was essentially disabled

SteveMacenski commented on nav2_waypoint_follower/plugins/wait_at_waypoint_arrival.cpp:
```diff
@@ -0,0 +1,100 @@
+    node->get_logger(), "Arrived at %i'th waypoint, sleeping for %i milliseconds",
+    curr_waypoint_index,
+    waypoint_pause_duration_);
+  RCLCPP_INFO(
```
I don't think this pose INFO is necessary

jediofgever replied:
we have set `-Werror` in nav2_package() macro, if I don't make use of pose, I get `[-Werror=unused-parameter]` on compilation. Though I agree it is not a very useful INFO

SteveMacenski replied:
@jediofgever you can get around that by commenting out that parameter but keep the signature, e.g.
`const int & /*curr_waypoint_index*/`

SteveMacenski commented on nav2_waypoint_follower/src/waypoint_follower.cpp:
```diff
@@ -23,14 +23,18 @@
 
   declare_parameter("stop_on_failure", true);
   declare_parameter("loop_rate", 20);
+  declare_parameter("task_executor_at_waypoint_plugin", default_task_executor_at_waypoint_id_);
```
You can just set the default plugin here, you don't need to store it as a class variable. Please remove the `default_XYZ_` variables and just put them in the declares.

SteveMacenski commented on nav2_waypoint_follower/src/waypoint_follower.cpp:
```diff
@@ -64,6 +70,28 @@ WaypointFollower::on_configure(const rclcpp_lifecycle::State & /*state*/)
+    task_executor_at_waypoint_type_ = nav2_util::get_plugin_type_param(
+      this,
+      task_executor_at_waypoint_id_);
+    task_executor_at_waypoint_ = task_executor_at_waypoint_loader_.createUniqueInstance(
```
`createUniqueInstance` returns a unique not a shared pointer, the `task_executor_at_waypoint_` variable is a shared pointer I believe

SteveMacenski reviewed (changes requested):
Really, really good first draft. I'm actually really impressed that you picked up on our plugin styling on your first PR in the project!

Just some stuff to clean up but overall I have no substantive issues.

jediofgever commented:
Thanks for the constructive feedback. I think I touched all the points on the requested changes. I am willing to add the tutorial to website as we have all needed code/classes. However I had like to start that once this is merged.

SteveMacenski commented on nav2_waypoint_follower/plugins/wait_at_waypoint.cpp:
```diff
@@ -0,0 +1,87 @@
+  node->get_parameter(
+    plugin_name + ".enabled",
+    is_enabled_);
+  if (!waypoint_pause_duration_) {
```
Check if 0, using a boolean operator on an int technically works, but that's really unclear and open for misinterpretation

SteveMacenski commented on doc/parameters/param_list.md:
```diff
@@ -479,6 +479,16 @@ When `planner_plugins` parameter is not overridden, the following default plugin
 | loop_rate | 20 | Rate to check for results from current navigation task |
+| waypoint_task_executor_plugin | `waypoint_task_executor` | Name of plugin to be loaded for executing waypoint tasks.|
+
+## waypoint_task_executor plugin
```
I think this is `WaitAtWaypoint` not `waypoint_task_executor`, since this is for a specific implementation of a waypoint task executor.

jediofgever replied:
Please refer here; https://github.com/ros-planning/navigation2/pull/1993/commits/4ad6f4dc2887dbdaf3d5aac557f351b74539af59#r499310248
I think it should be ok there ?

SteveMacenski replied:
No. `waypoint_task_executor` was correct and you didn't change the header name.

SteveMacenski commented on nav2_waypoint_follower/include/nav2_waypoint_follower/plugins/wait_at_waypoint.hpp:
```diff
@@ -0,0 +1,77 @@
+ */
+class WaitAtWaypoint : public nav2_core::WaypointTaskExecutor
+{
+private:
```
Move everything under `private` to the bottom. We usually do `public` functions, then `protected` variables. We want `protected` in case someone wants to subclass from this and wants to access those variables/.

SteveMacenski commented on nav2_waypoint_follower/plugins/wait_at_waypoint.cpp:
```diff
@@ -0,0 +1,87 @@
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    plugin_name + ".enabled",
+    rclcpp::ParameterValue(false));
```
Default this to true

SteveMacenski commented on nav2_waypoint_follower/plugins/wait_at_waypoint.cpp:
```diff
@@ -0,0 +1,87 @@
+      node->get_logger(),
+      "Waypoint pause duration is set to zero, disabling task executor plugin.");
+  }
+  if (!is_enabled_) {
```
This should be `else if` so that it only mentions disabling one time with the right context on why it was disabled

SteveMacenski commented on nav2_waypoint_follower/plugins/wait_at_waypoint.cpp:
```diff
@@ -0,0 +1,87 @@
+    logger_, "Arrived at %i'th waypoint, sleeping for %i milliseconds",
+    curr_waypoint_index,
+    waypoint_pause_duration_);
+  RCLCPP_INFO(
```
Remove INFO and you just need to do `/*curr_pose*/` in the function signature

SteveMacenski commented on nav2_waypoint_follower/src/waypoint_follower.cpp:
```diff
@@ -183,8 +210,12 @@ WaypointFollower::followWaypoints()
+        get_logger(), "Succeeded processing waypoint %i, processing waypoint task execution",
+        goal_index);
+      auto node = shared_from_this();
+      waypoint_task_executor_->processAtWaypoint(goal->poses[goal_index], goal_index);
```
Oh here's a good idea, we should change this API so that `processAtWaypoint` can return a `bool` and we can check if its `false` to also treat it as a fail checking on `stop_on_failure_` for whether to exit or continue to the next one

jediofgever replied:
nice idea, I changed the return type of `processAtWaypoint` to `bool`, but if the plugin is disabled or `waypoint_pause_duration` is set to zero, we will always have to return `false` from `processAtWaypoint` call, right ? that will make `stop_on_failure_` to be depending on the plugin to be enabled and active, not sure if that is behavior want ?

SteveMacenski replied:
If it fails to process & `stop_on_failure_` is true, we should exit with failure. 

I think maybe just not enabled returns true since its not a "failure", its just not turned on.

jediofgever replied:
I will address this after we confirm whether there will be a follow up PR

SteveMacenski commented on nav2_waypoint_follower/src/waypoint_follower.cpp:
```diff
@@ -64,6 +72,25 @@ WaypointFollower::on_configure(const rclcpp_lifecycle::State & /*state*/)
     get_node_waitables_interface(),
     "FollowWaypoints", std::bind(&WaypointFollower::followWaypoints, this), false);
 
+  nav2_util::declare_parameter_if_not_declared(
```
Add this in the constructor with the other declares.

SteveMacenski reviewed (changes requested):
Just some smaller clean up

SteveMacenski commented:
Also this PR needs to target `main` not `foxy-devel`

jediofgever commented:
> Also this PR needs to target `main` not `foxy-devel`

sure but `main` and `foxy-devel` seems to be diverged a lot, my branch is  `jediofgever:foxy-devel` and if I target it to `ros-planning:main` wouldn't that introduce some unwanted things here ? 

Followed your suggestions on docs branch and made another commit to address points. It will still need changes but priority now should be on how to merge this pull into `main`.  

To me it seems that, I will need to add the latest changes in this pull(which you have already reviewed) to my fork (`main`), and open a new(unfourtunately) pull request which targets to `ros-planning:main  `
what do you think ?

SteveMacenski commented:
I don't know, but we do not accept new feature development to a targeted distribution, it must be in `main`. You may have some conflicts to resolve  but I doubt it, you're messing with code that hasn't changed in a long time. 

We also need that docs PR to be submitted as well before this can be merged. I want to make sure that we keep things up to date.

SteveMacenski commented:
@jediofgever any update? That last thing with stop on failure is the last thing (other than going to main) for merging this!

jediofgever commented:
sorry I was away for a local holiday. Will update here soon.

SteveMacenski commented:
No worries, enjoy your vacation :palm_tree:

naiveHobo commented:
As much as I like the concept this PR introduces, isn't the whole idea of executing a behavior after reaching a waypoint better suited to be executed as part of the behavior tree? A very simple way of sleeping after reaching a waypoint would be by adding a `Wait` BT action node in sequence with the existing default behavior tree. A more nav2 way of solving it would be to implement waypoint follower itself as a BT (#1928) and adding complex behaviors to this BT. Thoughts?

jediofgever commented on doc/parameters/param_list.md:
```diff
@@ -479,7 +479,7 @@ When `planner_plugins` parameter is not overridden, the following default plugin
 | stop_on_failure | true | Whether to fail action task if a single waypoint fails. If false, will continue to next waypoint. |
 | loop_rate | 20 | Rate to check for results from current navigation task |
-| waypoint_task_executor_plugin | `waypoint_task_executor` | Name of plugin to be loaded for executing waypoint tasks.|
+| waypoint_task_executor_plugin | `WaitAtWaypoint` | Name of plugin to be loaded for executing waypoint tasks.|
```
@SteveMacenski, I think it is resolved here ? `waypoint_task_executor`==> `WaitAtWaypoint`

jediofgever commented:
It seems hard to elegantly change target branch to main with current status, 
there is a lot of conflicting files, It was my mistake to target `foxy-devel` at first place, let me fix that by placing latest changes to `main` and submit a follow up PR. This will make things look easier to understand and cleaner.

SteveMacenski commented:
> isn't the whole idea of executing a behavior after reaching a waypoint better suited to be executed as part of the behavior tree

That's one option, but sometimes that's overkill. Sometimes you just need a robot to follow some points and do something at each point, and in that case this could be a more streamline method. Plus if we wanted to add custom behavior to the WP follower this would be required, so I like the idea of making it a framework vs patching for every single possible thing someone might want to do at a waypoint for testing.

@jediofgever this PR has become somewhat of a mess, there's a ton of stuff in here that should _not_ be in here from incorrect rebasing or conflicting branches. I'm not sure what you did, but please revert those changes.

jediofgever commented:
> this PR has become somewhat of a mess

Agreed, but no worries I will handle it
